portalnet/src/utp_controller.rs

```diff
@@ -15,33 +25,253 @@ use crate::discovery::UtpEnr;
+        utp_socket: Arc<UtpSocket<UtpEnr>>,
+        metrics: OverlayMetricsReporter,
+    ) -> Self {
         Self {
             utp_socket,
             inbound_utp_transfer_semaphore: Arc::new(Semaphore::new(utp_transfer_limit)),
             outbound_utp_transfer_semaphore: Arc::new(Semaphore::new(utp_transfer_limit)),
+            metrics,
+        }
+    }
+
+    /// Connect with a peer given the connection id, and return the data received from the peer.
+    pub async fn connect_inbound_stream(
+        &self,
+        connection_id: u16,
+        peer: Enr,
+        trace: Option<QueryTrace>,
+    ) -> Result<Vec<u8>, OverlayRequestError> {
+        self.metrics
+            .report_utp_active_inc(UtpDirectionLabel::Inbound);
+        let cid = utp_rs::cid::ConnectionId {
+            recv: connection_id,
+            send: connection_id.wrapping_add(1),
+            peer: UtpEnr(peer),
+        };
+        let mut stream = match self.connect_with_cid(cid.clone(), *UTP_CONN_CFG).await {
+            Ok(stream) => stream,
+            Err(err) => {
+                self.metrics.report_utp_outcome(
+                    UtpDirectionLabel::Inbound,
+                    UtpOutcomeLabel::FailedConnection,
+                );
+                debug!(
+                    %err,
+                    cid.send,
+                    cid.recv,
+                    peer = ?cid.peer.client(),
+                    "Unable to establish uTP conn based on Content response",
+                );
+                return Err(OverlayRequestError::ContentNotFound {
+                    message:
+                        "Unable to locate content on the network: unable to establish utp conn"
+                            .to_string(),
+                    utp: true,
+                    trace,
+                });
+            }
+        };
+
+        let mut data = vec![];
+        if let Err(err) = stream.read_to_eof(&mut data).await {
+            self.metrics
+                .report_utp_outcome(UtpDirectionLabel::Inbound, UtpOutcomeLabel::FailedDataTx);
+            debug!(%err, cid.send, cid.recv, peer = ?cid.peer.client(), "error reading data from uTP stream, while handling a FindContent request.");
+            return Err(OverlayRequestError::ContentNotFound {
+                message:
+                    "Unable to locate content on the network: error reading data from utp stream"
+                        .to_string(),
+                utp: true,
+                trace,
+            });
+        }
+
+        // report utp tx as successful, even if we go on to fail to process the
+        // payload
+        self.metrics
+            .report_utp_outcome(UtpDirectionLabel::Inbound, UtpOutcomeLabel::Success);
+        Ok(data)
+    }
+
+    /// Connect with a peer given the connection id, and transfer the data to the peer.
+    pub async fn connect_outbound_stream(&self, cid: ConnectionId<UtpEnr>, data: Vec<u8>) -> bool {
```

Comment:
Can't the logic in `connect_outbound_stream` and `accept_outbound_stream` be condensed? there is a ton of duplicate code

Reply:
I don't think so... `connect_outbound_stream` establishes the stream via `connect_with_cid` unlike `accept_outbound_stream` which establishes the stream via `accept_with_cid`. Plus they have different `debug!` logs while doing so, and they return different values. The common logic between `connect_outbound_stream` & `accept_outbound_stream` has already been encapsulated into `send_utp_content`, which they both already use..

Reply:
I will take a shot at it then